platform/android/src/test/render_test_runner.cpp

```diff
@@ -41,17 +42,178 @@ void Log::platformRecord(EventSeverity severity, const std::string& msg) {
 
 } // namespace mbgl
 
+namespace {
+std::string jstring2string(JNIEnv* env, jstring jStr) {
+    if (!jStr) {
+        return "";
+    }
+    const jclass stringClass = env->GetObjectClass(jStr);
+    const jmethodID getBytes = env->GetMethodID(stringClass, "getBytes", "(Ljava/lang/String;)[B");
+    const jbyteArray stringJbytes = (jbyteArray)env->CallObjectMethod(jStr, getBytes, env->NewStringUTF("UTF-8"));
+
+    size_t length = (size_t)env->GetArrayLength(stringJbytes);
+    jbyte* pBytes = env->GetByteArrayElements(stringJbytes, NULL);
+
+    std::string ret = std::string((char*)pBytes, length);
+    env->ReleaseByteArrayElements(stringJbytes, pBytes, JNI_ABORT);
+
+    env->DeleteLocalRef(stringJbytes);
+    env->DeleteLocalRef(stringClass);
+    return ret;
+}
+
+bool copyFile(JNIEnv* env,
+              AAssetManager* assetManager,
+              const std::string& filePath,
+              const std::string& destinationPath,
+              const std::string& fileName) {
+    jclass fileClass = env->FindClass("java/io/File");
+    jmethodID fileCtor = env->GetMethodID(fileClass, "<init>", "(Ljava/lang/String;Ljava/lang/String;)V");
+    jmethodID fileExists = env->GetMethodID(fileClass, "exists", "()Z");
+
+    jstring destination = env->NewStringUTF(destinationPath.c_str());
+    jstring file = env->NewStringUTF(fileName.c_str());
+    jobject fileToCopy = env->NewObject(fileClass, fileCtor, destination, file);
+    bool stateOk = true;
+    if (env->CallBooleanMethod(fileToCopy, fileExists)) {
+        mbgl::Log::Warning(mbgl::Event::General, "File '%s' already exists", filePath.c_str());
+    } else {
+        AAsset* fileAsset = AAssetManager_open(assetManager, fileName.c_str(), AASSET_MODE_BUFFER);
+        const void* fileData = AAsset_getBuffer(fileAsset);
+        const off_t fileLen = AAsset_getLength(fileAsset);
+
+        FILE* newFile = std::fopen(filePath.c_str(), "w+");
+        stateOk = newFile != NULL;
+        if (!stateOk) {
+            mbgl::Log::Warning(mbgl::Event::General, "Failed to create new file entry %s", fileName.c_str());
+        } else {
+            auto res = std::fwrite(fileData, sizeof(char), fileLen, newFile);
+            if (fileLen != res) {
+                mbgl::Log::Warning(
+                    mbgl::Event::General, "Failed to generate file entry %s from assets", fileName.c_str());
+            }
+        }
+        std::fclose(newFile);
+        AAsset_close(fileAsset);
```

Comment:
what if an exception is thrown in the code above, the files will remain opened, right?